modules/api/VcenterConnection.py

```diff
@@ -0,0 +1,152 @@
+            return False
+        try:
+            self.api.CurrentTime()
+            return True
+        except (vim.fault.NotAuthenticated, http.client.RemoteDisconnected):
+            return False
+
+    def disconnect(self) -> None:
+        """
+        Disconnect from vCenter. Does not matter if there is a connection.
+        """
+        if self.api is not None:
+            Disconnect(self.api)
+
+    def _get_obj(self, root, vim_type):
+        """
+        Creates a ContainterView and returns it. Can be used to fetch all
+        hosts or vms in one turn.
+
+        :param root: The root where to start. eg vcenter.content.root
+        :param vim_type: vimtype eg vim.HostSystem or vim.vm
+        :return: ContainerView
+        """
+
+        container = self.api.content.viewManager.CreateContainerView(
+            root, vim_type, True)
+        view = container.view
+        container.Destroy()
+        return view
+
+    @staticmethod
+    def _create_filter_spec(esxi_hosts, prop):
+        """
+        Build a filter spec (some kind of sql_query) for the property
+        collector.
+        See community samples
+        https://github.com/vmware/pyvmomi-community-samples/blob
+        /bc14f63065aa360ceca0cca477a8b271d582a090/samples/filter_vms.py
+
+        :param esxi_hosts: a ContainerView of vim.HostSystem (Get with get_obj
+        :param prop: a single string or list of strings. Properties to
+        collect. Eg name, overallStatus...
+        :return: the created filter spec
+        """
+
+        if not isinstance(prop, list):
+            prop = [prop]
+
+        objSpecs = []
+        for host in esxi_hosts:
+            objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=host)
+            objSpecs.append(objSpec)
+        filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+        filterSpec.objectSet = objSpecs
+        propSet = vmodl.query.PropertyCollector.PropertySpec(all=False)
+        propSet.type = vim.HostSystem
+        propSet.pathSet = prop
+        filterSpec.propSet = [propSet]
+        return filterSpec
+
+    def get_esxi_overall_stats(self) -> List[Host]:
+        """
+        Get the overall status of all esxi-hosts from vcenter.
+        Status can be green, yellow or red.
+
+        :return: a list of interfaces.Host
+        """
+
+        self.login()
+        esxi = self._get_obj(self.api.content.rootFolder, [vim.HostSystem])
+        pc = self.api.content.propertyCollector
+        filter_spec = self._create_filter_spec(
+            esxi, ['overallStatus', 'name'])
+        options = vmodl.query.PropertyCollector.RetrieveOptions()
+        result = pc.RetrievePropertiesEx([filter_spec], options)
+        res = []
+        vcenter = Vcenter(name=self.host, address=self.host)
+        state_map = {
+            'green': 2, 'yellow': 1, 'red': 0,
+        }
+        for item in result.objects:
+            name = [v.val for v in item.propSet if v.name == 'name'][0]

```

Comment:
are we sure, to have an element in this list? If it wouldn't be filled we have an `IndexError`

Reply:
https://code.vmware.com/apis/968/vsphere > Types > HostSystem says;

it inherits `name` and `overall status` from _managed entity_. I never experienced a key error but I could add exception handling if you want.

Reply:
Just wanted to raise this. If it is always filled, all good.

Reply:
nevermind, I wrote a check procedure